third/nmh/man/msgchk.man
.\"
.\" %nmhwarning%
.\" $Id: msgchk.man,v 1.1.1.1 1999-02-07 18:14:21 danw Exp $
.\"
.\" include the -mh macro file
.so %etcdir%/tmac.h
.\"
.TH MSGCHK %manext1% MH.6.8 [%nmhversion%]
.SH NAME
msgchk \- check for messages
.SH SYNOPSIS
.in +.5i
.ti -.5i
msgchk
\%[\-date] \%[\-nodate]
\%[\-notify\ all/mail/nomail]
.br
\%[\-nonotify\ all/mail/nomail]
.br
%nmhbeginpop%
\%[\-host\ hostname]
\%[\-user\ username]
.br
%nmhendpop%
\%[users\ ...]
\%[\-version]
\%[\-help]
.in -.5i
.SH DESCRIPTION
The \fImsgchk\fR program checks all known mail drops for mail waiting
for you.  For those drops which have mail for you, \fImsgchk\fR will
indicate if it believes that you have seen the mail in question before.

The `\-notify\ type' switch indicates under what circumstances
\fImsgchk\fR should produce a message.  The default is `\-notify\ all'
which says that \fImsgchk\fR should always report the status of the
users maildrop.  Other values for `type' include `mail' which says that
\fImsgchk\fR should report the status of waiting mail; and, `nomail'
which says that \fImsgchk\fR should report the status of empty maildrops.
The `\-nonotify\ type' switch has the inverted sense, so
`\-nonotify\ all' directs \fImsgchk\fR to never report the status of
maildrops.  This is useful if the user wishes to check \fImsgchk\fR's
exit status.  A non\-zero exit status indicates that mail was \fBnot\fR
waiting for at least one of the indicated users.

If \fImsgchk\fR produces output, then the `\-date' switch directs
\fImsgchk\fR to print out the last date mail was read, if this can
be determined.
%nmhbeginpop%
 
.Uh "Using POP"
\fImsgchk\fR will normally check all the local mail drops, but if
the option \*(lqpophost:\*(rq is set in the mts configuration file
\*(lqmts.conf\*(rq, or if the `\-host\ hostname' switch is given,
\fImsgchk\fR will query this POP service host as to the status of
mail waiting.

The default is for \fImsgchk\fR to assume that your account name
on the POP server is the same as your current username.  To specify
a different username, use the `\-user\ username' switch.

When using POP, you will normally need to type the password for
your account on the POP server, in order to retrieve your messages.
It is possible to automate this process by creating a \*(lq.netrc\*(rq
file containing your login account information for this POP server.
For each POP server, this file should have a line of the following
form.  Replace the words mypopserver, mylogin, and mypassword with
your own account information.

machine mypopserver login mylogin password mypassword

This \*(lq.netrc\*(rq file should be owned and readable only by
you.

For debugging purposes, there is also a switch `\-snoop', which will
allow you to watch the POP transaction take place between you and the
POP server.
%nmhendpop%
.Fi
^$HOME/\&.mh\(ruprofile~^The user profile
^%etcdir%/mts.conf~^nmh mts configuration file
^%mailspool%/$USER~^Location of mail drop
.Pr
None
.Sa
inc(1)
.De
`user' defaults to the current user
.Ds
`\-date'
.Ds
`\-notify\ all'
.Co
None
.En


doc/news
This file is intended to keep track of user-visible changes to the
release, for the construction of the Athena release notes.

Changes from release 8.0 to release 8.1:

	* Athena version numbers are now dotted triplets, e.g. 8.1.0
	  or 8.1.12.  Paths which include all or part of the Athena
	  version number now take the form "pathname-version",
	  e.g. "srvd-8.1" or "dash.menus-8.1"

	* In release 8.0, running "xmkmf" with the default path would
	  generate a Makefile which looked for its include files and
	  libraries in /mit/x11, which could be one of several
	  lockers.  Now it is preferrable to use the vendor X
	  libraries, so this bit of glue is going away.  There will be
	  no /usr/athena/bin/xmkmf any more, nor any symlinks in
	  /usr/athena pointing to the x11 locker.

	  Users can still do something like "add -f x11r5" and then
	  "xmkmf", and things should work, if the x11r5 locker is
	  built for their platform.

	* Athena bitmaps have moved to /usr/athena/include/X11/bitmaps.
	  Previously, they were in /usr/athena/lib/X11/bitmaps before
	  because of the /mit/x11 glue.

	* The symlink /bin/csh -> /bin/athena/tcsh has gone away.  The
	  update to 8.0 will fix /bin/csh references in /etc/passwd
	  and /etc/passwd.local to refer to /bin/athena/tcsh instead.

	* emacs18 has gone away.

	* AFS includes and libraries are installed in
	  /usr/afsws/include and /usr/afsws/lib.

	* wcl includes and libraries are installed in /usr/athena.

	* The basic krb5 headers and libraries are installed in
	  /usr/athena.

	* emacs has been upgraded from 19.30 to 19.34.  There should
	  be no significant user-visible changes.

	* RCS has been upgraded from version 4 to version 5.7.

	* x3270 has been upgraded from version 1.2 to version 3.1.1.3.

	* sysinfo is now available under Irix.

	* /bin/athena/tcsh now has Hesiod support enabled, which means
	  that "~lockername" will work in a pathname in most cases.
	  (It won't work for users with local passwd information who
	  don't have their lockers attached.)

	* For improved ease of debugging, many binaries are installed
	  with symbol tables, and compiled without optimization.
	  Therefore, the contents of /bin/athena and /usr/athena/bin
	  may be somewhat larger, and the binaries may consume more
	  CPU cycles, although it's unlikely that anyone will notice
	  the speed difference.


third/nmh/uip/msgchk.c

/*
 * msgchk.c -- check for mail
 *
 * $Id: msgchk.c,v 1.1.1.1 1999-02-07 18:14:15 danw Exp $
 */

#include <h/mh.h>
#include <zotnet/mts/mts.h>
#include <zotnet/tws/tws.h>
#include <pwd.h>

#ifdef POP
# include <h/popsbr.h>
#endif

#ifdef HESIOD
# include <hesiod.h>
#endif

#ifndef	POP
# define POPminc(a) (a)
#else
# define POPminc(a)  0
#endif

#ifndef	RPOP
# define RPOPminc(a) (a)
#else
# define RPOPminc(a)  0
#endif

#ifndef	APOP
# define APOPminc(a) (a)
#else
# define APOPminc(a)  0
#endif

static struct swit switches[] = {
#define	DATESW                   0
    { "date", 0 },
#define	NDATESW                  1
    { "nodate", 0 },
#define	NOTESW                   2
    { "notify type", 0 },
#define	NNOTESW                  3
    { "nonotify type", 0 },
#define	HOSTSW                   4
    { "host hostname", POPminc (-4) },
#define	USERSW                   5
    { "user username", POPminc (-4) },
#define	APOPSW                   6
    { "apop", APOPminc (-4) },
#define	NAPOPSW                  7
    { "noapop", APOPminc (-6) },
#define	RPOPSW                   8
    { "rpop", RPOPminc (-4) },
#define	NRPOPSW                  9
    { "norpop", RPOPminc (-6) },
#define VERSIONSW               10
    { "version", 0 },
#define	HELPSW                  11
    { "help", 4 },
#define SNOOPSW                 12
    { "snoop", -5 },
    { NULL, 0 }
};

/*
 * Maximum numbers of users we can check (plus
 * one for the NULL vector at the end).
 */
#define MAXVEC  51

#define	NT_NONE	0x0
#define	NT_MAIL	0x1
#define	NT_NMAI	0x2
#define	NT_ALL	(NT_MAIL | NT_NMAI)

#define	NONEOK	0x0
#define	UUCPOLD	0x1
#define	UUCPNEW	0x2
#define	UUCPOK	(UUCPOLD | UUCPNEW)
#define	MMDFOLD	0x4
#define	MMDFNEW	0x8
#define	MMDFOK	(MMDFOLD | MMDFNEW)


/*
 * static prototypes
 */
static int donote (char *, int);
static int checkmail (char *, char *, int, int, int);

#ifdef POP
static int remotemail (char *, char *, int, int, int, int);
#endif


int
main (int argc, char **argv)
{
    int datesw = 1, notifysw = NT_ALL;
    int rpop, status = 0;
    int snoop = 0, vecp = 0;
    uid_t uid;
    char *cp, *host = NULL, *user, buf[BUFSIZ];
    char **argp, **arguments, *vec[MAXVEC];
    struct passwd *pw;

#ifdef HESIOD
    struct hes_postoffice *po;
    char *tmphost;
#endif

#ifdef LOCALE
    setlocale(LC_ALL, "");
#endif
    invo_name = r1bindex (argv[0], '/');

    /* read user profile/context */
    context_read();

    mts_init (invo_name);
    uid = getuid ();
    user = getusername();

    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

#ifdef POP
    if ((cp = getenv ("MHPOPDEBUG")) && *cp)
	snoop++;
#endif

#ifdef KPOP
    rpop = 1;
#else
    rpop = 0;
#endif

    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
		case AMBIGSW: 
		    ambigsw (cp, switches);
		    done (1);
		case UNKWNSW: 
		    adios (NULL, "-%s unknown", cp);

		case HELPSW: 
		    snprintf (buf, sizeof(buf), "%s [switches] [users ...]",
			invo_name);
		    print_help (buf, switches, 1);
		    done (1);
		case VERSIONSW:
		    print_version(invo_name);
		    done (1);

		case DATESW:
		    datesw++;
		    continue;
		case NDATESW:
		    datesw = 0;
		    continue;

		case NOTESW:
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    notifysw |= donote (cp, 1);
		    continue;
		case NNOTESW:
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    notifysw &= ~donote (cp, 0);
		    continue;

		case HOSTSW: 
		    if (!(host = *argp++) || *host == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    continue;
		case USERSW: 
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    if (vecp >= MAXVEC-1)
			adios (NULL, "you can only check %d users at a time", MAXVEC-1);
		    else
			vec[vecp++] = cp;
		    continue;

		case APOPSW: 
		    rpop = -1;
		    continue;
		case NAPOPSW:
		    rpop = 0;
		    continue;

		case RPOPSW: 
		    rpop = 1;
		    continue;
		case NRPOPSW: 
		    rpop = 0;
		    continue;

		case SNOOPSW:
		    snoop++;
		    continue;
	    }
	}
	if (vecp >= MAXVEC-1)
	    adios (NULL, "you can only check %d users at a time", MAXVEC-1);
	else
	    vec[vecp++] = cp;
    }

#ifdef POP
    /*
     * If -host is not specified by user
     */
    if (!host || !*host) {
# ifdef HESIOD
	/*
	 * Scheme is:
	 *        use MAILHOST environment variable if present,
	 *  else try Hesiod.
	 *  If that fails, use the default (if any)
	 *  provided by mts.conf in mts_init()
	 */
	if ((tmphost = getenv("MAILHOST")) != NULL)
	    pophost = tmphost;
	else if ((po = hes_getmailhost(vecp ? vec[0] : user)) != NULL &&
		strcmp(po->po_type, "POP") == 0)
	    pophost = po->po_host;
# endif /* HESIOD */
	/*
	 * If "pophost" is specified in mts.conf,
	 * use it as default value.
	 */
	if (pophost && *pophost)
	    host = pophost;
    }
    if (!host || !*host)
	host = NULL;
    if (!host || rpop <= 0)
	setuid (uid);
#endif /* POP */

    if (vecp != 0)
	vec[vecp] = NULL;

#ifdef POP
    if (host) {
	if (vecp == 0) {
	    status = remotemail (host, user, rpop, notifysw, 1, snoop);
	} else {
	    for (vecp = 0; vec[vecp]; vecp++)
		status += remotemail (host, vec[vecp], rpop, notifysw, 0, snoop);
	}
    } else {
#endif /* POP */

    if (vecp == 0) {
	char *home;

	home = (uid = geteuid()) ? home = getenv ("HOME") : NULL;
	if (home == NULL) {
	    pw = getpwnam (user);
	    if (pw == NULL)
		adios (NULL, "unable to get information about user");
	    if (home == NULL)
		home = pw->pw_dir;
	}
	status = checkmail (user, home, datesw, notifysw, 1);
    } else {
	for (vecp = 0; vec[vecp]; vecp++) {
	    if ((pw = getpwnam (vec[vecp])))
		status += checkmail (pw->pw_name, pw->pw_dir, datesw, notifysw, 0);
	    else
		advise (NULL, "no such user as %s", vec[vecp]);
	}
    }
#ifdef POP
    }		/* host == NULL */
#endif

    done (status);
}


static struct swit ntswitches[] = {
#define	NALLSW     0
    { "all", 0 },
#define	NMAISW     1
    { "mail", 0 },
#define	NNMAISW	   2
    { "nomail", 0 },
    { NULL, 0 }
};


static int
donote (char *cp, int ntflag)
{
    switch (smatch (cp, ntswitches)) {
	case AMBIGSW: 
	    ambigsw (cp, ntswitches);
	    done (1);
	case UNKWNSW: 
	    adios (NULL, "-%snotify %s unknown", ntflag ? "" : "no", cp);

	case NALLSW: 
	    return NT_ALL;
	case NMAISW: 
	    return NT_MAIL;
	case NNMAISW: 
	    return NT_NMAI;
    }
}


static int
checkmail (char *user, char *home, int datesw, int notifysw, int personal)
{
    int mf, status;
    char buffer[BUFSIZ];
    struct stat st;

    snprintf (buffer, sizeof(buffer), "%s/%s", mmdfldir[0] ? mmdfldir : home, mmdflfil[0] ? mmdflfil : user);
    if (datesw) {
	st.st_size = 0;
	st.st_atime = st.st_mtime = 0;
    }
    mf = (stat (buffer, &st) == NOTOK || st.st_size == 0) ? NONEOK
	: st.st_atime <= st.st_mtime ? MMDFNEW : MMDFOLD;

    if ((mf & UUCPOK) || (mf & MMDFOK)) {
	if (notifysw & NT_MAIL) {
	    printf (personal ? "You have " : "%s has ", user);
	    if (mf & UUCPOK)
		printf ("%s old-style bell", mf & UUCPOLD ? "old" : "new");
	    if ((mf & UUCPOK) && (mf & MMDFOK))
		printf (" and ");
	    if (mf & MMDFOK)
		printf ("%s%s", mf & MMDFOLD ? "old" : "new",
			mf & UUCPOK ? " Internet" : "");
	    printf (" mail waiting");
	} else {
	    notifysw = 0;
	}
	status = 0;
    }
    else {
	if (notifysw & NT_NMAI)
	    printf (personal ? "You don't %s%s" : "%s doesn't %s",
		    personal ? "" : user, "have any mail waiting");
	else
	    notifysw = 0;

	status = 1;
    }

    if (notifysw)
	if (datesw && st.st_atime)
	    printf ("; last read on %s", dtime (&st.st_atime, 1));
    if (notifysw)
	printf ("\n");

    return status;
}


#ifdef POP
extern char response[];

static int
remotemail (char *host, char *user, int rpop, int notifysw, int personal, int snoop)
{
    int nmsgs, nbytes, status;
    char *pass = NULL;

    if (user == NULL)
	user = getusername ();
    if (rpop > 0)
	pass = getusername ();
    else
	ruserpass (host, &user, &pass);

    /* open the POP connection */
    if (pop_init (host, user, pass, snoop, rpop) == NOTOK
	    || pop_stat (&nmsgs, &nbytes) == NOTOK	/* check for messages  */
	    || pop_quit () == NOTOK) {			/* quit POP connection */
	advise (NULL, "%s", response);
	return 1;
    }

    if (nmsgs) {
	if (notifysw & NT_MAIL) {
	    printf (personal ? "You have " : "%s has ", user);
	    printf ("%d message%s (%d bytes)",
		    nmsgs, nmsgs != 1 ? "s" : "", nbytes);
	}
	else
	    notifysw = 0;

	status = 0;
    } else {
	if (notifysw & NT_NMAI)
	    printf (personal ? "You don't %s%s" : "%s doesn't %s",
		    personal ? "" : user, "have any mail waiting");
	else
	    notifysw = 0;
	status = 1;
    }
    if (notifysw)
	printf (" on %s\n", host);

    return status;
}
#endif /* POP */
